File: bindings/react-native/android/src/main/java/com/offlineprotocol/TransportManager.kt
package com.offlineprotocol

/**
 * Transport lifecycle states
 */
enum class TransportState {
    UNAVAILABLE,
    AVAILABLE,
    STARTING,
    RUNNING,
    STOPPING,
    STOPPED
}

/**
 * Listener interface for transport events
 */
interface TransportManagerListener {
    /**
     * Called when transport state changes
     */
    fun onTransportStateChanged(manager: TransportManager, state: TransportState)
    
    /**
     * Called when transport encounters an error
     */
    fun onTransportError(manager: TransportManager, error: Throwable)

    /**
     * Called when the transport emits a diagnostic message
     */
    fun onTransportDiagnostic(manager: TransportManager, level: String, message: String, context: Map<String, Any?>)
}

/**
 * Base interface for all transport implementations
 * This allows for extensible transport system supporting BLE, WiFi Direct, and Internet
 */
interface TransportManager {
    /**
     * Unique transport identifier
     */
    val transportId: String
    
    /**
     * Human-readable transport name
     */
    val transportName: String
    
    /**
     * Current transport state
     */
    val state: TransportState
    
    /**
     * Listener for transport events
     */
    var listener: TransportManagerListener?
    
    /**
     * Checks if the transport is available on this device
     * @return true if transport hardware/capabilities are available
     */
    fun isAvailable(): Boolean
    
    /**
     * Starts the transport
     * @throws TransportException if start fails
     *
     * ## Threading
     *
     * An implementation must not run UniFFI calls on the caller's thread.
     * Every call into the core serialises on one global protocol mutex, held
     * across MLS work and platform keystore access, so a lifecycle method that
     * calls the protocol inline hands those waits to whoever called it — and
     * the callers here include the app's main thread. Confine the work
     * instead; [TransportConfinement] is what the shipped transports use.
     */
    fun start()

    /**
     * Stops the transport
     *
     * Same threading obligation as [start].
     */
    fun stop()

    /**
     * Pauses the transport (for background mode)
     *
     * The default delegates to [stop] and therefore inherits the caller's
     * thread. Every transport in this module overrides it, so nothing reaches
     * this body today — but a new transport that does not override is exactly
     * how lifecycle FFI creeps back onto main. Override both, or confine
     * inside [start] and [stop].
     *
     * Confining does not mean posting and returning. This must have *taken
     * effect* by the time it returns: `OfflineProtocolModule.pause` pauses
     * every transport and then the core, and that order is what bounds the
     * window in which a paused transport can still re-enter UniFFI. An
     * implementation that hands the work to its handler and returns has paused
     * nothing yet, so the core can pause underneath it. Wait on the
     * confinement — the module's caller is a React Native background thread,
     * where [TransportConfinement.runSync] does not bound the wait.
     */
    fun pause() {
        // Default implementation: stop the transport
        stop()
    }

    /**
     * Resumes the transport from paused state
     *
     * Same caveat as [pause].
     */
    fun resume() {
        // Default implementation: try to start the transport
        try {
            start()
        } catch (e: Exception) {
            // Ignore errors on resume
        }
    }
}

/**
 * Transport-specific exceptions
 */
sealed class TransportException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class NotAvailable(reason: String) : TransportException("Transport not available: $reason")
    class PermissionDenied(permission: String) : TransportException("Permission denied: $permission")
    class StartFailed(reason: String, cause: Throwable? = null) : TransportException("Failed to start transport: $reason", cause)
    class AlreadyRunning : TransportException("Transport is already running")
    class NotRunning : TransportException("Transport is not running")
    class InvalidState(message: String) : TransportException("Invalid state: $message")
    class PlatformError(cause: Throwable) : TransportException("Platform error: ${cause.message}", cause)
}

File: bindings/react-native/ios/TransportManager.swift
//
// TransportManager.swift
// OfflineProtocol
//
// Transport manager protocol for extensible transport implementations
//

import Foundation

/// Transport lifecycle states
public enum TransportState {
    case unavailable
    case available
    case starting
    case running
    case stopping
    case stopped
}

/// Delegate for transport events
public protocol TransportManagerDelegate: AnyObject {
    /// Called when transport state changes
    func transportManager(_ manager: TransportManager, didChangeState state: TransportState)
    
    /// Called when transport encounters an error
    func transportManager(_ manager: TransportManager, didEncounterError error: Error)
    
    /// Called when transport metrics are updated
    func transportManager(_ manager: TransportManager, didUpdateMetrics metrics: [String: Any])

    /// Called when the transport emits a diagnostic message
    func transportManager(_ manager: TransportManager, didEmitDiagnostic level: String, message: String, context: [String: Any])
}

/// Base protocol for all transport implementations
/// This allows for extensible transport system supporting BLE, WiFi Direct, and Internet
public protocol TransportManager: AnyObject {
    /// Unique transport identifier
    var transportId: String { get }
    
    /// Human-readable transport name
    var transportName: String { get }
    
    /// Current transport state
    var state: TransportState { get }
    
    /// Delegate for transport events
    var delegate: TransportManagerDelegate? { get set }
    
    /// Checks if the transport is available on this device
    /// - Returns: true if transport hardware/capabilities are available
    func isAvailable() -> Bool
    
    /// Starts the transport
    /// - Throws: TransportError if start fails
    func start() throws
    
    /// Stops the transport
    func stop()
    
    /// Pauses the transport (for background mode)
    func pause()
    
    /// Resumes the transport from paused state
    func resume()
    
    /// Gets current transport metrics
    /// - Returns: Dictionary of metric name to value
    func getMetrics() -> [String: Any]
}

/// Transport-specific errors
public enum TransportError: LocalizedError {
    case notAvailable(String)
    case permissionDenied(String)
    case startFailed(String)
    case alreadyRunning
    case notRunning
    case invalidState(String)
    case platformError(Error)
    
    public var errorDescription: String? {
        switch self {
        case .notAvailable(let reason):
            return "Transport not available: \(reason)"
        case .permissionDenied(let permission):
            return "Permission denied: \(permission)"
        case .startFailed(let reason):
            return "Failed to start transport: \(reason)"
        case .alreadyRunning:
            return "Transport is already running"
        case .notRunning:
            return "Transport is not running"
        case .invalidState(let message):
            return "Invalid state: \(message)"
        case .platformError(let error):
            return "Platform error: \(error.localizedDescription)"
        }
    }
}

/// Extension for default implementations
extension TransportManager {
    public func pause() {
        // Default implementation: stop the transport
        stop()
    }
    
    public func resume() {
        // Default implementation: try to start the transport
        try? start()
    }
    
    public func getMetrics() -> [String: Any] {
        // Default implementation: return empty metrics
        return [:]
    }
}

